### PR TITLE
Feature/315 implementing portfolio system

### PR DESCRIPTION
**Pull request by DaveBouman**
## Description

There are still some things to do. Mainly some tests and a updating the resourceResults. But I already wanted some input on what to change before the code freeze.

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [X] This change requires a documentation update

## Checklist

-   [X] My code follows the style guidelines of this project
-   [X] I did not update API Controllers, if I did, I added/changed Postman tests
-   [X] I have performed a self-review of my own code
-   [X] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [X] New and existing unit tests pass locally with my changes
-   [ ] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.
These steps will be used during release testing.

1. Go to api
2. Try to make new portfolio and portfolio items

## [Link to issue](https://github.com/DigitalExcellence/dex-backend/issues/315)

Closes: 315

### CODE REVIEW

File: Data/4_Data.csproj [hunk 1]
```diff
@@ -11,6 +11,15 @@
...
   </ItemGroup>
+
+  <ItemGroup>
+    <Compile Remove="Migrations\20201202082834_Initportfolioitems.cs" />
+    <Compile Remove="Migrations\20201202082834_Initportfolioitems.Designer.cs" />
+    <Compile Remove="Migrations\20201202091420_CreateListOfPortfolioItemsinPortfolioTable.cs" />
+    <Compile Remove="Migrations\20201202091420_CreateListOfPortfolioItemsinPortfolioTable.Designer.cs" />
+    <Compile Remove="Migrations\20210105075132_portfolioItemNoLinkedTable.cs" />
+    <Compile Remove="Migrations\20210105075132_portfolioItemNoLinkedTable.Designer.cs" />
+  </ItemGroup>
```
Comment (Ruby77) [inline Data/4_Data.csproj:22]: I think this can be removed.

File: Services/Services/PortfolioService.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        public override void Add(Portfolio entity)
+        {
+            base.Add(entity);
+        }
+
+        public override Task AddAsync(Portfolio entity)
+        {
+            return base.AddAsync(entity);
+        }
```
Comment (Ruby77) [inline Services/Services/PortfolioService.cs:49]: What is the added value by overriding a method and calling the base method again?

**Review by Ruby77 — CHANGES_REQUESTED** [review]
The code looks good. I won't dive too deep yet, because I understand a lot can get changed 😄 . I noticed a few things that looked weird to me. But overall it looks really good. Let me know when everything is finished so I can give the final, in-depth review :)

File: Data/4_Data.csproj (hunk 1 above)
Comment (DaveBouman) [reply Data/4_Data.csproj]: I have removed it

File: Services/Services/PortfolioService.cs (hunk 2 above)
Comment (DaveBouman) [reply Services/Services/PortfolioService.cs]: I changed it.

File: Services/Services/PortfolioItemService.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        public override string ToString()
+        {
+            return base.ToString();
+        }
+
+        public override void Update(PortfolioItem entity)
+        {
+            base.Update(entity);
+        }
```
Comment (Ruby77) [inline Services/Services/PortfolioItemService.cs:110]: I think this can be removed? Overriding a method and calling the base method doesn't add functionality I think.

File: Services/Services/PortfolioService.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        Task<PortfolioItem> IPortfolioService.DeletePortfolioItem(PortfolioItem portfolioItem)
+        {
+            throw new NotImplementedException();
+        }
+
+        Task<PortfolioItem> IPortfolioService.UpdatePortfolioItem(PortfolioItem portfolioItem)
+        {
+            throw new NotImplementedException();
+        }
```
Comment (Ruby77) [inline Services/Services/PortfolioService.cs:54]: Is there a reason why certain methods are not implemented yet?

File: Services/Services/ProjectService.cs [hunk 5]
```diff
@@ -72,7 +72,7 @@ public override void Add(Project entity)
 
         public override void Update(Project entity)
         {
-            // Sanitize description before executing default behaviour.
+            /// Sanitze description before executing default behaviour.
```
Comment (Ruby77) [inline Services/Services/ProjectService.cs:75]: Why the change? I think the first version was the correct grammar :)

File: API/Resources/PortfolioItemResource.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+{
+    /// <summary>
+    /// The View model of a portfolio item
+    /// </summary>
+    public class PortfolioItemResource
+    {
+        ///// <summary>
+        ///// This sets or gets the portfolio associated with the item
+        ///// </summary>
+        //public int PortfolioId { get; set; }
```
Comment (Ruby77) [inline API/Resources/PortfolioItemResource.cs:17]: Remove commented code

File: API/Resources/PortfolioItemResource.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        /// <summary>
+        /// this sets or gets the content
+        /// </summary>
+        public string Content { get; set; }
+
+        ///// <summary>
+        ///// this sets or gets the content
+        ///// </summary>
+        //public int Project { get; set; }
+    }
```
Comment (Ruby77) [inline API/Resources/PortfolioItemResource.cs:38]: Remove commented code

**Review by Ruby77 — CHANGES_REQUESTED** [review]
I won't be able to pull your branch and test every functionality yet for the upcoming days. I will try to do this at the end of the week. Although the code looks good, made some small notes.

**Comment by Maxvanhattum** [comment]
While running the postman tests I get the following exception: 
https://gyazo.com/2d0420501c4f1b2245c93aaebe39644c
Am I the only one with this problem?

It happens because multiple elements are present, and SingleOrDefault doesnt like that. If not intended, FirstOrDefault can be used.

**Comment by DaveBouman** [comment]
> While running the postman tests I get the following exception:
> https://gyazo.com/2d0420501c4f1b2245c93aaebe39644c
> Am I the only one with this problem?
> 
> It happens because multiple elements are present, and SingleOrDefault doesnt like that. If not intended, FirstOrDefault can be used.

I will look into it. If I will get the same errors

File: Repositories/PortfolioRepository .cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        /// <summary>
+        /// Find the portfolio async by portfolio id
+        /// </summary>
+        /// <param name="id">The identifier.</param>
+        /// <returns>
+        public override async Task<Portfolio> FindAsync(int id)
+        {
+            Portfolio portfolio = await GetDbSet<Portfolio>()
+           .Where(s => s.Id == id)
```
Comment (Maxvanhattum) [inline Repositories/PortfolioRepository .cs:57]: You need to include Portfolio items here.

File: API/Configuration/MappingProfile.cs [hunk 9]
```diff
@@ -62,14 +63,31 @@ public MappingProfile()
...
                 .ForMember(q => q.Description, opt => opt.MapFrom(p => p.Project.Description))
                 .ForAllOtherMembers(o => o.Ignore());
 
+            CreateMap<Portfolio, PortfolioResourceResult>()
+               .ForMember(q => q.Id, opt => opt.MapFrom(p => p.Id))
+               .ForMember(q => q.Template, opt => opt.MapFrom(p => p.Template))
+               .ForMember(q => q.Title, opt => opt.MapFrom(p => p.Title))
+               .ForMember(q => q.UserId, opt => opt.MapFrom(p => p.User.Id))
+               .ForMember(q => q.UserName, opt => opt.MapFrom(p => p.User.Name))
+               .ForMember(q => q.PortfolioItem, opt => opt.MapFrom(p => p.PortfolioItem))
```
Comment (Maxvanhattum) [inline API/Configuration/MappingProfile.cs:80]: This will fail since Portfolio.Portfolio item is a different list type then that of PortfolioResourceResult. You will notice this once the projectitems are included, they are empty as of now.

File: API/Controllers/PortfolioController.cs [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+        [ProducesResponseType(typeof(PortfolioItemResourceResult), (int) HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.Unauthorized)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.NotFound)]
+        public async Task<IActionResult> CreatePortfolioItemAsync(int portfolioId, int projectId, PortfolioItemResource portfolioItemResource)
+        {
+            User user = await HttpContext.GetContextUser(userService).ConfigureAwait(false);
+
+            Portfolio portfolio = await portfolioService.FindAsync(portfolioId).ConfigureAwait(false);
+            Project project = await projectService.FindAsync(projectId).ConfigureAwait(false);
+            PortfolioItem portfolioItem = mapper.Map<PortfolioItemResource, PortfolioItem>(portfolioItemResource);
```
Comment (Maxvanhattum) [inline API/Controllers/PortfolioController.cs:327]: This fails to map the project id, leaving it at 0.

**Review by Maxvanhattum — CHANGES_REQUESTED** [review]
Hey hey, it isn't quite working for me. I can create a portfolio, and I can create portfolio items. However, after I created items, and I retrieve the portfolio, the items aren't linked. When retrieving the portfolio items manually, it says it is linked to portfolio with Id 0. The project is also id 0. Look at the comments I left on how to fix these issues, as for now however this PR is broken.

Also take a look at all your comments, quite a lot are missing the proper mark up. Moreover, the responses the routes provide are quite bloated, everything is populated, so when requesting a portfolio it loads the user, his likes, the projects, their likes etc. I don't think this is necessary .

As a final note, I see no tests are made, I think for this PR it would be best if they'd be included, not in a separate issue/PR. This is just my opinion tho, however it would have caught that this PR is basically not functioning as expected,  now it depended on human/manual testing.

File: API/Resources/LimitedPortfolioItemResourceResult .cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+using Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace API.Resources
+{
+    /// <summary>
+    /// viewmodel to return to frontend
```
[the remark this answers is not in the document]
Comment (DaveBouman) [reply API/Resources/LimitedPortfolioItemResourceResult .cs]: I removed the resource result

File: API/Controllers/PortfolioController.cs [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+    [Route("api/[controller]")]
+    [ApiController]
+    public class PortfolioController : ControllerBase
+    {
+        private readonly IMapper mapper;
+        private readonly IPortfolioService portfolioService;
+        private readonly IPortfolioItemService portfolioItemService;
+        private readonly IProjectService projectService;
+        private readonly IUserService userService;
+        private readonly IAuthorizationHelper authorizationHelper;
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:34]: The authorization helper is not used. Although, it could give added value. Consider if the data officer should be able to make requests in name of a user in their institute (in my eyes this could absolutely work)

File: API/Controllers/PortfolioController.cs [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+        /// <response code="200">This endpoint returns the portfolio with the specified portfolio id.</response>
+        /// <response code="400">The 400 Bad Request status code is returned when the portfolio id is invalid.</response>
+        /// <response code="404">The 404 Not Found status code is returned when the portfolio with the specified id could not be found.</response>
+        [HttpGet("{portfolioId}")]
+        [ProducesResponseType(typeof(PortfolioResourceResult), (int) HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.BadRequest)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.NotFound)]
+        public async Task<IActionResult> GetPortfolio(int portfolioId)
+        {
+            if(portfolioId < 0)
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:76]: I think, an id should not be zero. (If you agree, you should also change the detail property from the problem object.)

```suggestion
            if(portfolioId <= 0)
```

File: API/Controllers/PortfolioController.cs [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+                };
+                return BadRequest(problem);
+            }
+
+            Portfolio portfolio = mapper.Map<PortfolioResource, Portfolio>(portfolioResource);
+
+            try
+            {
+                portfolio.User = user;
+                portfolioService.Add(portfolio);
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:144]: I think you can use the overloaded async version here.
```suggestion
                await portfolioService.AddAsync(portfolio);
```

File: API/Controllers/PortfolioController.cs [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+                    Title = "Failed to update portfolio.",
+                    Detail = "The specified portfolio could not be found in the database.",
+                    Instance = "65229A71-FDD8-4887-BBCD-3D4258C5D329"
+                };
+                return NotFound(problem);
+            }
+
+
+
+            if(!(portfolio.User.Id == user.Id))
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:195]:
```suggestion
            if(portfolio.User.Id != user.Id)
```

File: API/Controllers/PortfolioController.cs [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+        public async Task<IActionResult> DeletePortfolio(int portfolioId)
+        {
+            Portfolio portfolio = await portfolioService.FindAsync(portfolioId).ConfigureAwait(false);
+            User user = await HttpContext.GetContextUser(userService).ConfigureAwait(false);
+
+            if(portfolio.User.Id != user.Id)
+                return Forbid();
+
+
+            if(portfolio == null)
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:233]: The way you structured this code, it not completely correct. The check in the if statement (line 233) always returns false. The reason is: line 229 tries to access a property in portfolio. So if this would be null, it already throws an error. I hope this makes sense :) (please let me know if it does not).

So a solution could be to switch the order of both checks.

File: API/Controllers/PortfolioController.cs [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+                ProblemDetails problem = new ProblemDetails
+                {
+                    Title = "Failed to delete portfolio.",
+                    Detail = "The specified portfolio could not be found in the database.",
+                    Instance = "CF5D3E72-6B98-4828-BDBB-E829BDED297D"
+                };
+                return NotFound(problem);
+            }
+
+            if(!(portfolio.User.Id == user.Id))
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:244]:
```suggestion
            if(portfolio.User.Id != user.Id)
```

File: API/Controllers/PortfolioController.cs [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+        /// <returns>This method returns the user resource result.</returns>
+        /// <response code="200">This endpoint returns the portfolio with the specified portfolio item id.</response>
+        /// <response code="400">The 400 Bad Request status code is returned when the portfolio item id is invalid.</response>
+        /// <response code="404">The 404 Not Found status code is returned when the portfolio item with the specified id could not be found.</response>
+        [HttpGet("item/{portfolioItemId}")]
+        [ProducesResponseType(typeof(PortfolioItemResourceResult), (int) HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.NotFound)]
+        public async Task<IActionResult> GetPortfolioItem(int portfolioItemId)
+        {
+            if(portfolioItemId < 0)
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:278]:
```suggestion
            if(portfolioItemId <= 0)
```

File: API/Controllers/PortfolioController.cs [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+            else
+            {
+                portfolioItem.ProjectId = 0;
+            }
+
+            try
+            {
+                portfolioItem.ProjectId = project.Id;
+                portfolioItem.PortfolioId = portfolio.Id;
+                portfolioItemService.Add(portfolioItem);
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:382]:
```suggestion
                await portfolioItemService.AddAsync(portfolioItem);
```

File: API/Controllers/PortfolioController.cs [hunk 20]
```diff
@@ -0,0 +1,498 @@
...
+                    Title = "Failed to update portfolio item.",
+                    Detail = "The specified portfolio could not be found in the database.",
+                    Instance = "8F2CDC1B-7015-4B72-B91B-78FD839DDEF1"
+                };
+                return NotFound(problem);
+            }
+
+            User user = await HttpContext.GetContextUser(userService).ConfigureAwait(false);
+
+            if(!(portfolio.User.Id == user.Id))
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:433]:
```suggestion
            if(portfolio.User.Id != user.Id)
```

File: API/Controllers/PortfolioController.cs [hunk 21]
```diff
@@ -0,0 +1,498 @@
...
+                ProblemDetails problem = new ProblemDetails
+                {
+                    Title = "Failed to delete portfolio item.",
+                    Detail = "The specified portfolio could not be found in the database.",
+                    Instance = "618CCDA0-A8E2-4B8E-B393-661D96A22585"
+                };
+                return NotFound(problem);
+            }
+
+            if(!(portfolio.User.Id == user.Id))
```
Comment (Ruby77) [inline API/Controllers/PortfolioController.cs:479]:
```suggestion
            if(portfolio.User.Id != user.Id)
```

**Review by Ruby77 — CHANGES_REQUESTED** [review]
Hey, looks good. Good work! I have some remarks and a question about your implementation.

You can specify for your portfolioItem the position, I think it's the order which the elements should get rendered on the frontend. I noticed you are able to set multiple times the same position, how does this work? 

Thanks @DaveBouman for you work!

File: API/Controllers/PortfolioController.cs (hunk 16 above)
Comment (DaveBouman) [reply API/Controllers/PortfolioController.cs:233]: I changed the postion of those two. I Also deleted the check if user id is the portfolio user id. Because it double checked the same parameters

**Comment by Ruby77** [comment]
Hey @DaveBouman, not all the feedback is processed yet. Is this something you still would like to work on, or do you want someone else to continue working on this (because of your internship that will start)? Good luck with your internship!

**Comment by niraymak** [comment]
Closed due inactivity. This issue needs some more research in how we want to implement it. Branch stays open till we know what solution we will choose.
